### PR TITLE
config/DockerImplementation

### PR DESCRIPTION
**Pull request by Josecisneros001**
This pr includes a Dockerfile to define the docker image and also includes a Makefile with the most useful docker commands for those who do not have much experience, here are some examples:
- make build // To create the image
- make create // To create container
- make shell  // To fire up a bash session
- make down // To stop container
- make up // To start container

And there are more useful commands, later an entry would be made to the wiki explaining these commands.

The Container is created with some specific configurations:
- catkin_home folder is a shared volume between the container and the host, this to edit files from outside and see the changes reflected inside.
- same network configuration as the host.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+# ----------------------------------------------------------------------
+#  Development Melodic
+# ----------------------------------------------------------------------
+
+#: Builds a Docker image from the current Dockerfile file
+melodic.build:
+	@docker build -t ros:melodic -f docker/melodic/Dockerfile .
+melodic.speech.build:
+	@docker build -t ros:melodic-speech -f docker/melodic/Dockerfile.speech .
```
Comment (AuroTB) [inline Makefile:51]: Note: I would really like for the documentation to mention that we have to define the context with -f

**Review by AuroTB — CHANGES_REQUESTED** [review]
Speech docker build gives me some errors when running it with the `make melodic.speech.build` and when I tried to build it manually.
image


There were also some warnings with some dependencies, but other than that all other docker files run perfectly fine, good job!

The make file you've created is really great!
Please eliminate all noetic stuff.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,135 @@
+# ----------------------------------------------------------------------
+#  Development Noetic
+# ----------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply Makefile]: Good catch!

File: .gitignore [hunk 3]
```diff
@@ -14,6 +14,14 @@
 /catkin_home/src/.rosinstall.bak
 
 ### DeepSpeech && Speech API ###
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/build
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/dist
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/kenlm
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/*.egg-info
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/openfst*
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/openfst*
+/catkin_home/src/action_selectors/scripts/DeepSpeech/decoders/swig/swig_decoders.py
```
[the remark this answers is not in the document]
Comment (Josecisneros001) [reply .gitignore:23]: Yes they are, swig is a package that is build with cmake everytime. It creates a bunch of files.

File: Makefile (hunk 1 above)
Comment (Josecisneros001) [reply Makefile:9]: I will consider this once i start working in the wiki entry.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (AuroTB) [reply .gitignore:22]: cool! thanks for putting it in the git ignore then

**Comment by AYM1607** [comment]
I was unable to build the container image locally in MacOS. Here are the logs in case they're useful.

[logs.txt](https://github.com/RoBorregos/Robocup-Home/files/6063685/logs.txt)

**Review by AuroTB — APPROVED** [review]
Hello! After running everything the only detail would be adding to a README that this docker implementation only works for Ubuntu18 and below, because of a [dependency](https://www.isitdownrightnow.com/openslr.org.html) using ipv4.

Also add a TODO comment in the MakeFile.
Great work!

**Pull request merged**